experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,662 @@
+    fundamental::Vector{MPolyDecRingElem}
+    primary::Vector{MPolyDecRingElem}
+    secondary::Vector{MPolyDecRingElem}
+    
+    reynolds_operator::Function
+    
+    NullConeIdeal::MPolyIdeal
+    
+    #Invariant ring of reductive group G (in representation R), no other input.
+    function InvariantRing(R::RepresentationReductiveGroup) #here G already contains information n and rep_mat
+        z = new()
+        if isdefined(R, :weights)
+            n = length(weights(R))
+            z.field = field(group(R))
+            super_ring, __ = graded_polynomial_ring(field(group(R)), "X"=>1:n)
+            z.poly_ring = super_ring
+            z.group = group(R)
+            z.representation = R
+            #z.fundamental = torus_invariants_fast(weights(R), ring)
+            return z
+        end
+        n = ncols(R.rep_mat)
+        z.representation = R
+        z.group = R.group
+        G = z.group
+        z.field = G.field
+        z.poly_ring, __ = graded_polynomial_ring(G.field, "X" => 1:n)
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+    
+    #to compute invariant ring ring^G where G is the reductive group of R. 
+    function InvariantRing(R::RepresentationReductiveGroup, ring::MPolyDecRing)
+        n = ncols(R.rep_mat)
+        n == ngens(ring) || error("The given polynomial ring is not compatible.")
+        z = new()
+        if isdefined(R, :weights)
+            #dosomething
+        end
+        z.representation = R
+        z.group = R.group
+        G = R.group
+        z.field = G.field
+        z.poly_ring = ring
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+end
+
+invariant_ring(R::RepresentationReductiveGroup) = InvariantRing(R)
+invariant_ring(ring::MPolyDecRing, R::RepresentationReductiveGroup) = InvariantRing(R, ring)
+null_cone_ideal(R::InvariantRing) = R.NullConeIdeal
+poly_ring(R::InvariantRing) = R.poly_ring
+
+function fundamental_invariants(z::InvariantRing)
+    if isdefined(z, :fundamental)
+        return z.fundamental
+    else
+        R = z.representation
+        if group(group(R))[1] == :torus && isdefined(R, :weights)
+            z.fundamental = torus_invariants_fast(weights(R), poly_ring(z))
+            return z.fundamental
+        end
+        I, M = proj_of_image_ideal(R.group, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(R.group, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, R.group, z.poly_ring, M, I, z.reynolds_operator)
+        return z.fundamental
+    end
+end
+
+function Base.show(io::IO, R::InvariantRing) 
+    io = AbstractAlgebra.pretty(io)
+    println(io, "Invariant Ring of")
+    show(io, R.poly_ring)
+    print(io, AbstractAlgebra.Indent(),  " under group action of ", R.group.group[1], R.group.group[2])
+    print(io, AbstractAlgebra.Dedent())
+end
+
+#computing the graph Gamma from Derksens paper
+function image_ideal(G::ReductiveGroup, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
```

Comment:
You are needlessly restricting the second argument to a specific type which may cause issues down the road for some ring/group combos. Just relax it (then the code also looks simpler)
```suggestion
function image_ideal(G::ReductiveGroup, rep_mat::MatElem)
```
of course this applies to multiple other places in this file